src/Generating/FinishGen.cpp

```diff
@@ -775,36 +830,26 @@ void cFinishGenSprinkleFoliage::GenFinish(cChunkDesc & a_ChunkDesc)
 					{
 						a_ChunkDesc.SetBlockTypeMeta(x, ++Top, z, E_BLOCK_TALL_GRASS, E_META_TALL_GRASS_GRASS);
 					}
-					else if (TryAddSugarcane(a_ChunkDesc, x, Top, z))
-					{
-						++Top;
-					}
 					else if ((val1 > 0.5) && (val2 < -0.5))
 					{
 						a_ChunkDesc.SetBlockTypeMeta(x, ++Top, z, E_BLOCK_PUMPKIN, static_cast<int>(val3 * 8) % 4);
 					}
+					else
+					{
+						TryAddSugarcane(a_ChunkDesc, x, Top, z);
+					}
 					break;
 				}  // case E_BLOCK_GRASS
 
 				case E_BLOCK_SAND:
 				{
-					int y = Top + 1;
-					if (
-						(x > 0) && (x < cChunkDef::Width - 1) &&
-						(z > 0) && (z < cChunkDef::Width - 1) &&
-						(val1 + val2 > 0.5f) &&
-						(a_ChunkDesc.GetBlockType(x + 1, y, z)     == E_BLOCK_AIR) &&
-						(a_ChunkDesc.GetBlockType(x - 1, y, z)     == E_BLOCK_AIR) &&
-						(a_ChunkDesc.GetBlockType(x,     y, z + 1) == E_BLOCK_AIR) &&
-						(a_ChunkDesc.GetBlockType(x,     y, z - 1) == E_BLOCK_AIR) &&
-						IsDesertVariant(a_ChunkDesc.GetBiome(x, z))
-					)
+					if (val1 + val2 > 0.5f)
```

Comment:
This change will make sugar cane rarer in non-desert biomes.  You should keep the `IsDesertVariant` in the if.